### PR TITLE
Bug Fix in Image Generator - Now Allows FCN's to be used

### PR DESCRIPTION
**Pull request by joeyearsley**
I needed FCN's so have altered this preprocessing step, I can also make a gist for the tutorials if anyone else prefers to learn through a simple example.

This does not allow for no sizes to be set if max poolings are used. In that case, the sizes can be none-square but have to be multiple of 2^m - where m is the number of max pooling layers.

In the future it might be possible to auto-resize to the nearest factor, but another variable will need adding to ensure we can disentangle the model from the preprocessing.

Made minor changes to allow for FCN's:
  - Target Size is now a tuple of (None, None) as only this can propagate through the directory iterator
  - Altered batch_x to be created in the loop, as you should only use a batch size of 1 for FCN of shape (None, None, Channels)
  - Allowed the directory iterator to use the folder it is passed if it finds no subfolders for classes (Usual use case in FCN)

Allowed BGR to be specified (for when using caffe converted models).

### CODE REVIEW

File: keras/preprocessing/image.py [hunk 1]
```diff
@@ -191,7 +191,7 @@ def flip_axis(x, axis):
     return x
 
 
-def array_to_img(x, dim_ordering='default', scale=True):
+def array_to_img(x, dim_ordering='default', scale=True, color_mode='rgb'):
```
Comment (fchollet) [inline keras/preprocessing/image.py]: If you add an argument you should document it in the docstring.

File: keras/preprocessing/image.py [hunk 2]
```diff
@@ -241,7 +243,7 @@ def array_to_img(x, dim_ordering='default', scale=True):
         raise ValueError('Unsupported channel number: ', x.shape[2])
 
 
-def img_to_array(img, dim_ordering='default'):
+def img_to_array(img, dim_ordering='default', color_mode='rgb'):
```
Comment (fchollet) [inline keras/preprocessing/image.py]: Same.

Comment (joggino) [inline keras/preprocessing/image.py]: Fixed both, please tell me if the docstring doesn't suffice.

**Comment by ahundt** [comment]
@joeyearsley I'd appreciate an example. Also, https://github.com/aurora95/Keras-FCN has some good scripts for training FCNs

File: keras/preprocessing/image.py [hunk 3]
```diff
@@ -298,7 +308,7 @@ def load_img(path, grayscale=False, target_size=None):
         img = img.convert('L')
     else:  # Ensure 3 channel even when loaded image is grayscale
         img = img.convert('RGB')
-    if target_size:
+    if target_size[0] and target_size[1]:  # Can't pass None in from iterator, hence tuple
```
Comment (fchollet) [inline keras/preprocessing/image.py]: We can keep the default to `None` and convert to `(None, None)` here. At the very least, `target_size=None` should keep being supported (for backwards compatibility).

File: keras/preprocessing/image.py [hunk 4]
```diff
@@ -777,6 +787,9 @@ def __init__(self, directory, image_data_generator,
             for subdir in sorted(os.listdir(directory)):
                 if os.path.isdir(os.path.join(directory, subdir)):
                     classes.append(subdir)
+        # User passed in None and no sub-folders, hence FCN. N.B.: Pass in classes=['.'] for improvements.
```
Comment (fchollet) [inline keras/preprocessing/image.py]: I cannot quite understand this comment. Can you clarify?

Also: line too long.

**Comment by ahundt** [comment]
@fchollet I believe by FCN he means support for dense prediction tasks like image segmentation of which the FCN paper happens to be one implementation. 

This class of tasks is not yet well supported by Keras in the "best practices by default" sense in which Keras is designed. A [number](https://github.com/aurora95/Keras-FCN) [of](https://github.com/nicolov/segmentation_keras) [implementations](https://github.com/PavlosMelissinos/enet-keras/) have popped up to handle the problem. Examples of missing keras capabilities required for best practice dense prediction include:
- APIs need to apply augmentation (zoom, flip, translation) to both the input data and the labels
- Class weights do not work with segmentation predictions
- The keras loss functions don't support "2D" loss, such as comparing against a (height x width x num_classes) tensor
- no segmentation datasets are yet supported which I'm trying to alleviate in keras-contrib with [coco](https://github.com/farizrahman4u/keras-contrib/pull/81) and [pascalvoc](https://github.com/farizrahman4u/keras-contrib/pull/80). 

It would be nice if support for dense prediction could be added to the [requests for contributions](https://github.com/fchollet/keras/projects/1) so dense prediction improvements would qualify for direct inclusion in keras, rather than going through keras-contrib.

**Comment by fchollet** [comment]
@ahundt would you like to take over that project? It would involve:

- coming up with a list of features to support
- agreeing on a set of API changes and getting them approved (by me)
- implementing it, either by managing the work of other contributors or coding it yourself
- reviewing any PRs related to this goals (such as the present PR)

Yes, we can add to the 'request for contributions' section.

File: keras/preprocessing/image.py [hunk 5]
```diff
@@ -822,22 +835,30 @@ def next(self):
             index_array, current_index, current_batch_size = next(self.index_generator)
         # The transformation of images is not under thread lock
         # so it can be done in parallel
-        batch_x = np.zeros((current_batch_size,) + self.image_shape)
+        # If image_shape has None in position 1 it is intended to be used as a FCN
+        if self.image_shape[1]:
+            batch_x = np.zeros((current_batch_size,) + self.image_shape)
```
Comment (ahundt) [inline keras/preprocessing/image.py]: does this assume all the images have the same shape? 
If so, varying image dimensions should be handled, and some kind of target dimensions could be utilized to apply padding to the images so they are all the same size.

reference code:
https://github.com/aurora95/Keras-FCN/blob/370eb767df86f364d93eff9068a3edb0b194a18b/utils/SegDataGenerator.py#L221
https://github.com/nicolov/segmentation_keras/blob/master/utils/image_reader.py#L79

File: keras/preprocessing/image.py [hunk 6]
```diff
@@ -822,22 +835,30 @@ def next(self):
...
         for i, j in enumerate(index_array):
             fname = self.filenames[j]
             img = load_img(os.path.join(self.directory, fname),
                            grayscale=grayscale,
                            target_size=self.target_size)
-            x = img_to_array(img, dim_ordering=self.dim_ordering)
+            x = img_to_array(img, dim_ordering=self.dim_ordering, color_mode=self.color_mode)
+            # Allows shape to be None, but batch size must equal one.
+            if not self.image_shape[1] and i == 0:
+                if current_batch_size != 1:
```
Comment (ahundt) [inline keras/preprocessing/image.py]: Why must batch size be one? Just looking for clarification because I'm not 100% sure what this case is checking for because dense prediction tasks including FCNs can train with batch sizes >1.

File: tests/keras/preprocessing/test_image.py [hunk 7]
```diff
@@ -160,6 +160,33 @@ def test_directory_iterator(self):
         assert(sorted(dir_iterator.filenames) == sorted(filenames))
         shutil.rmtree(tmp_folder)
 
+    ''' Test for the directory iterator when we wish to use it for a FCN.'''
+    def test_directory_iterator_fcn(self):
```
Comment (ahundt) [inline tests/keras/preprocessing/test_image.py]: I suggest also checking if labels can be loaded from pngs without palette conversion as described in the following, perhaps an additonal color mode 'segmentation_label' would be appropriate:
https://github.com/nicolov/segmentation_keras/issues/14

Though perhaps something like that is out of scope for this pull request.

**Comment by ahundt** [comment]
@fchollet sure I can give organizing that project a shot, though it may depend on how busy my research gets this summer. The bullets from my previous comment are a start. Where should this discussion continue so this pull request doesn't need to be hijacked any more than I've already done?

**Comment by fchollet** [comment]
@ahundt please open an issue with the list of changes you would like to see, and we will create a 'request for contributions' entry that will link to the issue. When you have an API proposal ready for review, please ping me.
